ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -947,7 +947,7 @@ Pair<Boolean, String> canCBOHandleAst(ASTNode ast, QB qb, PreCboCtx cboCtx) {
     // Now check QB in more detail. canHandleQbForCbo returns null if query can
     // be handled.
     msg = CalcitePlanner.canHandleQbForCbo(queryProperties, conf, true, needToLogMessage);
-    if (msg == null) {
+    if (msg == null || msg.isEmpty()) {
       return Pair.of(true, msg);
     }
     msg = msg.substring(0, msg.length() - 2);
```

Comment:
Checking if `msg.isEmpty` is a reasonable fix to avoid the IOBE here. The code though seems very brittle and relies on various assumptions that are not specified anywhere. 

As an alternative fix I would propose to delete this line altogether and do the same a few lines above where the same pattern appears.
```suggestion
```
I don't think we need the extra complexity just to remove a space and semicolon (`; `) from a few messages/explains. I doubt anyone will care. WDYT?

Reply:
Moreover, I would argue that allowing `CalcitePlanner.canHandleQbForCbo(queryProperties, conf, true, needToLogMessage)` to return an empty `String` is not good either but we can discuss this separately under another JIRA.

Reply:
Oh okay. That makes sense. We could keep the space and semicolon. I wasn't sure why that was added. I looked around for other places if we did something similar and I couldn't find. So, we are consistent also that way.

Reply:
Yeah, that's right

Reply:
Also, some of the explain analyze queries would show something like this

`Plan not optimized by CBO because the statement has PTF; ` as opposed to 
`Plan not optimized by CBO because the statement has PTF
`
which should be fine I guess

Reply:
This will need some changes to the query output files as well. Also, since it is affecting the aesthetics for the user directly, I am now thinking whethe we should really remove the below line
`msg = msg.substring(0, msg.length() - 2);`. What do you say?

Reply:
I would likely either - move the substring to canHandleQbForCbo, so that method always returns a valid looking message or change canHandleQbForCbo to construct a List<String> of reasons and then use String.join with "; " (and remove the substring code).

I might also even change the method not to bother with taking in verbose and to always create and return error messages. This is on the error path and not some inner loop reading data, so the performance optimization of checking log levels, etc is in the camp of unneeded optimization making the code ugly.

Though I'd go with what @zabetak prefers since they are the primary reviewer.

Reply:
The current changes in the PR solve the IOBE, tests are green, and there is no change in behavior so I plan to merge this now.

The rest of the changes proposed by @jfsii are an improvement worth having and are inline with what I briefly mentioned previously:

> allowing `CalcitePlanner.canHandleQbForCbo(queryProperties, conf, true, needToLogMessage)` to return an empty `String` is not good

I would like to get these additional changes merged as well so let's log a new JIRA and I will review ASAP. If nobody takes it in the next few days I can also work on it.

Reply:
Great. Thanks for the review Stamatis!